src/roinput.rs

```diff
@@ -0,0 +1,856 @@
+use super::*;
+use ark_ff::PrimeField;
+use bitvec::{prelude::*, view::AsBits};
+
+pub trait Input: Copy {
+    fn to_roinput(self) -> ROInput;
+    // The domain string length must be <= 20 (TODO: Limit at compile time)
+    fn domain_string(self, network_id: NetworkId) -> &'static str;
+}
+
+pub struct ROInput {
+    pub fields: Vec<PallasField>,
+    pub bits: BitVec<Lsb0, u8>,
+}
+
+impl ROInput {
+    pub fn new() -> Self {
+        ROInput {
+            fields: vec![],
+            bits: BitVec::new(),
+        }
+    }
+
+    pub fn append_field(&mut self, f: PallasField) {
+        self.fields.push(f);
+    }
+
+    pub fn append_scalar(&mut self, s: PallasScalar) {
+        // mina scalars are 255 bytes
+        let bytes = s.to_bytes(); // TODO: Combine these two into one-liner
+        let bits = &bytes.as_bits::<Lsb0>()[..PallasScalar::size_in_bits()];
+        self.bits.extend(bits);
+    }
+
+    pub fn append_bit(&mut self, b: bool) {
+        self.bits.push(b);
+    }
+
+    pub fn append_bytes(&mut self, bytes: Vec<u8>) {
+        self.bits.extend_from_bitslice(bytes.as_bits::<Lsb0>());
+    }
+
+    pub fn append_u32(&mut self, x: u32) {
+        self.append_bytes(x.to_le_bytes().to_vec());
+    }
+
+    pub fn append_u64(&mut self, x: u64) {
+        self.append_bytes(x.to_le_bytes().to_vec());
+    }
+
+    pub fn to_bytes(&mut self) -> Vec<u8> {
+        let mut bits: BitVec<Lsb0, u8> = self.fields.iter().fold(BitVec::new(), |mut acc, fe| {
+            acc.extend_from_bitslice(
+                &fe.to_bytes().as_bits::<Lsb0>()[..PallasField::size_in_bits()],
+            );
+
+            acc
+        });
+
+        bits.extend(self.bits.iter());
+
+        return bits.as_raw_slice().to_vec();
+    }
+
+    pub fn to_fields(&mut self) -> Vec<PallasField> {
```

Comment:
This is an interesting point.

You could even get the same fields if, for example, the preamble didn't consume the entire last field element in length and then you appended `0` bits at the end or something.

Maybe could disable it after it's been used, so that you can call `to_bytes` or `to_fields` again, but you are no-longer able to modify it with the `append_` methods.

Reply:
I think that this can still happen even if we make ROInput one time.  There's nothing stopping them from creating a new one and setting one more zero bit and yet getting the same ROInput output when executing to_fields.  I think perhaps the ROInput output may require a length associated with it.  Leaving this for future work as it would not be compatible with current protocol.